### PR TITLE
Dynamic job configuration

### PR DESCRIPTION
**Pull request by zonArt**
Give users the possibility to perform dynamic job configuration based on concat fragment and use them to create jobs within jenkins-cli

### CODE REVIEW

File: manifests/job/present.pp [hunk 1]
```diff
@@ -7,7 +7,10 @@
 # Parameters:
 #
 #   config
-#     the content of the jenkins job config file (required)
+#     the content of the jenkins job config file (template based)
```
Comment (jhoblitt) [inline manifests/job/present.pp]: The parameter is just a string.  No templating is applied.

File: manifests/job/present.pp [hunk 2]
```diff
@@ -7,7 +7,10 @@
...
+#
+#   temp_config_path
```
Comment (jhoblitt) [inline manifests/job/present.pp]: How about removing the 'temp_' prefix from the parameter name?  I think the name will confuse end users.

File: manifests/job/present.pp [hunk 3]
```diff
@@ -29,12 +33,43 @@
   include jenkins::cli
   include jenkins::cli::reload
 
+  if $temp_config_path != undef and $config != undef {
```
Comment (jhoblitt) [inline manifests/job/present.pp]: I believe  this could be simplified to `if $temp_config_path and $config {`.

File: manifests/job/present.pp [hunk 4]
```diff
@@ -29,12 +33,43 @@
...
+  }
+
   if $jenkins::service_ensure == 'stopped' or $jenkins::service_ensure == false {
     fail('Management of Jenkins jobs requires \$jenkins::service_ensure to be set to \'running\'')
   }
 
   $jenkins_cli        = $jenkins::cli::cmd
-  $tmp_config_path    = "/tmp/${jobname}-config.xml"
+  if $config == undef {
+    $tmp_config_path = $temp_config_path
```
Comment (jhoblitt) [inline manifests/job/present.pp]: I feel that use two variable names that are so similar is asking for trouble down the road...

File: manifests/job/present.pp [hunk 5]
```diff
@@ -29,12 +33,43 @@
   include jenkins::cli
   include jenkins::cli::reload
 
+  if $temp_config_path != undef and $config != undef {
+    fail('You cannot set both $temp_config_path AND $config param, only one is required')
+  }
```
Comment (jhoblitt) [inline manifests/job/present.pp]: A test is also needed to  make sure one of config_path or config is defined.

**Comment by zonArt** [comment]
@jhoblitt Thanks for the comments, I'll take them into consideration and push an updated proposal, will try the testing in a third commit

**Comment by zonArt** [comment]
Any news on this one ? Is it possible to get it merge ?

**Comment by jhoblitt** [comment]
There still isn't any test coverage of the new `config_file` parameter.

**Comment by zonArt** [comment]
Ok, so if I understand that well, it means I need to write a new spec/defines/jenkins_job_present.rb with all the tests related to this define ? Is it the strategy to adopt (I'm sorry for the maybe stupid question but I'm totally newb at writing puppet tests)

**Comment by zonArt** [comment]
@jhoblitt, I made a test on config_file as requested, don't know if this is enough or not

**Comment by edomaur** [comment]
:+1:  seems fine now :-)

**Comment by edomaur** [comment]
:+1: seems fine now :-)

**Comment by zonArt** [comment]
What about this PR? I guess the "needs-tests" label could be removed and what about merging it?

**Comment by zonArt** [comment]
Beaker tests are now fixed thanks to @jhoblitt, could we now consider this as a merge candidate ?

**Comment by zonArt** [comment]
@jhoblitt Is it possible to get this merged before new conflict occurs ? If anything is missing or wrongly done, thanks to tell so that we can go further on with it

**Comment by edomaur** [comment]
+1

**Pull request merged**